### PR TITLE
Compile against ref assemblies

### PR DESCRIPTION
**Pull request by JunTaoLuo**
https://github.com/aspnet/AspNetCore-Internal/issues/1773

### CODE REVIEW

**Comment by JunTaoLuo** [comment]
There are lots of failures due to internal APIs which are no longer shared since non-public APIs are not included in the ref assemblies. One option is to include internal APIs in ref assemblies, but I ran into issues where GenAPI generated uncompilable code. The alternative would be to move the internal API to shared files. It's a little tedious but I'll start on this tomorrow if we can't find other solutions.

Thoughts @dougbu @wtgodbe ?

**Comment by dsplaisted** [comment]
@ericstj Any suggestions for building against reference assemblies but accessing internal APIs?

File: eng/GenAPI.exclusions.txt [hunk 1]
```diff
@@ -2,4 +2,11 @@
 T:Microsoft.AspNetCore.Components.RenderTree.RenderTreeFrame
 # Manually implemented - https://github.com/dotnet/arcade/issues/2066
 T:Microsoft.AspNetCore.Mvc.ApplicationModels.PageParameterModel
-T:Microsoft.AspNetCore.Mvc.ApplicationModels.PagePropertyModel
\ No newline at end of file
+T:Microsoft.AspNetCore.Mvc.ApplicationModels.PagePropertyModel
+# Manually implemented - Need to include internal setter
```
Comment (JunTaoLuo) [inline eng/GenAPI.exclusions.txt]: TODO: Followup

Maybe there's a way to do this without needing to explicitly opt out. Will follow up with experts but for now this works.

File: src/Mvc/Mvc.Core/src/Formatters/MediaTypeHeaderValues.cs [hunk 2]
```diff
@@ -16,6 +16,9 @@ internal static class MediaTypeHeaderValues
         public static readonly MediaTypeHeaderValue ApplicationAnyJsonSyntax
             = MediaTypeHeaderValue.Parse("application/*+json").CopyAsReadOnly();
 
+        public static readonly MediaTypeHeaderValue ApplicationJsonPatch
```
Comment (JunTaoLuo) [inline src/Mvc/Mvc.Core/src/Formatters/MediaTypeHeaderValues.cs]: Merging implementations: Mvc.NewtonsoftJson/src/MediaTypeHeaderValues.cs.

File: src/Mvc/Mvc.NewtonsoftJson/src/MediaTypeHeaderValues.cs [hunk 3]
```diff
@@ -1,22 +0,0 @@
-// Copyright (c) .NET Foundation. All rights reserved.
```
Comment (JunTaoLuo) [inline src/Mvc/Mvc.NewtonsoftJson/src/MediaTypeHeaderValues.cs]: Revert this change.

File: src/Mvc/Mvc.TagHelpers/src/Cache/CacheTagKey.cs [hunk 4]
```diff
@@ -5,6 +5,7 @@
 using System.Collections.Generic;
 using System.Globalization;
 using System.Text;
+using Microsoft.AspNetCore.Antiforgery;
```
Comment (JunTaoLuo) [inline src/Mvc/Mvc.TagHelpers/src/Cache/CacheTagKey.cs]: Revert

File: src/Mvc/Mvc.Core/src/Microsoft.AspNetCore.Mvc.Core.csproj [hunk 5]
```diff
@@ -48,6 +49,7 @@ Microsoft.AspNetCore.Mvc.RouteAttribute</Description>
   </ItemGroup>
 
   <ItemGroup Label="Sources packages">
+    <Reference Include="Microsoft.Extensions.HashCodeCombiner.Sources" />
```
Comment (dougbu) [inline src/Mvc/Mvc.Core/src/Microsoft.AspNetCore.Mvc.Core.csproj]: With correct `InternalsVisibleTo` attributes in ref/ assemblies, why are changes to this project necessary?

File: src/Mvc/Mvc.NewtonsoftJson/src/Microsoft.AspNetCore.Mvc.NewtonsoftJson.csproj [hunk 6]
```diff
@@ -21,7 +21,6 @@
   </ItemGroup>
 
   <ItemGroup>
-    <Compile Include="..\..\Mvc.Core\src\Formatters\ResponseContentTypeHelper.cs" />
```
Comment (dougbu) [inline src/Mvc/Mvc.NewtonsoftJson/src/Microsoft.AspNetCore.Mvc.NewtonsoftJson.csproj]: Is there a _new_ `InternalsVisibleTo` relationship that makes this source file unnecessary?

File: src/Mvc/Mvc.Razor/ref/Microsoft.AspNetCore.Mvc.Razor.netcoreapp3.0.cs [hunk 7]
```diff
@@ -290,14 +290,6 @@ public partial class ViewsFeature
         public System.Collections.Generic.IList<Microsoft.AspNetCore.Mvc.Razor.Compilation.CompiledViewDescriptor> ViewDescriptors { [System.Runtime.CompilerServices.CompilerGeneratedAttribute]get { throw null; } }
     }
 }
-namespace Microsoft.AspNetCore.Mvc.Razor.Infrastructure
```
Comment (dougbu) [inline src/Mvc/Mvc.Razor/ref/Microsoft.AspNetCore.Mvc.Razor.netcoreapp3.0.cs]: Since we're dealing w/ partial classes, I suggest (just suggest) instead removing the `public` members from the .Manual file.

File: src/Mvc/Mvc.Razor/src/Microsoft.AspNetCore.Mvc.Razor.csproj [hunk 8]
```diff
@@ -11,6 +11,8 @@
   </PropertyGroup>
 
   <ItemGroup>
+    <Compile Include="$(SharedSourceRoot)CryptographyAlgorithms\**\*.cs" />
```
Comment (dougbu) [inline src/Mvc/Mvc.Razor/src/Microsoft.AspNetCore.Mvc.Razor.csproj]: I'd rather we minimize the number of approaches we're taking to get things working. Please file a follow-up issue to go back to unshared `internal`s with `InternalsVisibleTo`.

File: src/SignalR/common/Shared/PipeWriterStream.cs [hunk 9]
```diff
@@ -1,12 +1,9 @@
-﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
-using System;
 using System.Buffers;
-using System.IO;
 using System.Threading;
 using System.Threading.Tasks;
-using Microsoft.AspNetCore.Internal;
```
Comment (dougbu) [inline src/SignalR/common/Shared/PipeWriterStream.cs]: Revert changes here unless a namespace is disappearing for some reason. Same for other `using ...Internal` cleanup.

And, if namespaces are disappearing, why?

File: src/DataProtection/Cryptography.Internal/ref/Directory.Build.props [hunk 10]
```diff
@@ -0,0 +1,8 @@
+<Project>
+  <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory)..\, Directory.Build.props))\Directory.Build.props" />
+
+  <PropertyGroup>
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
+    <NoWarn>$(NoWarn);CS0169</NoWarn>
```
[the remark this answers is not in the document]
Comment (JunTaoLuo) [reply src/DataProtection/Cryptography.Internal/ref/Directory.Build.props]: The CS0169 should not be needed if we remove the private APIs. AllowUnsafeBlocks is only needed in certain ref assemblies. I think I'll keep these files for now.

File: src/Mvc/Mvc.Razor/ref/Microsoft.AspNetCore.Mvc.Razor.netcoreapp3.0.cs (hunk 7 above)
Comment (JunTaoLuo) [reply src/Mvc/Mvc.Razor/ref/Microsoft.AspNetCore.Mvc.Razor.netcoreapp3.0.cs]: The reason why I needed to do this is that the public class here has an internal setter on a public property. I don't think we can separate that so I moved the whole type to Manual.cs.

File: src/SignalR/common/Shared/PipeWriterStream.cs (hunk 9 above)
Comment (JunTaoLuo) [reply src/SignalR/common/Shared/PipeWriterStream.cs]: These aren't used. The Microsoft.AspNetCore.Internal is no longer available since the internal APIs are not generated by default so this using must be removed. This originally was added to support an internal extension which has since been removed.

File: src/DataProtection/Cryptography.Internal/ref/Microsoft.AspNetCore.Cryptography.Internal.Manual.cs [hunk 11]
```diff
@@ -0,0 +1,422 @@
...
+[assembly: InternalsVisibleTo("Microsoft.AspNetCore.DataProtection.Extensions.Tests, PublicKey=0024000004800000940000000602000000240000525341310004000001000100f33a29044fa9d740c9b3213a93e57c84b472c84e0b8a0e1ae48e67a9f8f6de9d5f7f3d52ac23e48ac51801f1dc950abe901da34d2a9e3baadb141a17c77ef3c565dd5ee5054b91cf63bb3c6ab83f72ab3aafe93d0fc3c2348b764fafb0b1c0733de51459aeab46580384bf9d74c4e28164b7cde247f891ba07891c9d872ad2bb")]
+[assembly: InternalsVisibleTo("Microsoft.AspNetCore.DataProtection.Tests, PublicKey=0024000004800000940000000602000000240000525341310004000001000100f33a29044fa9d740c9b3213a93e57c84b472c84e0b8a0e1ae48e67a9f8f6de9d5f7f3d52ac23e48ac51801f1dc950abe901da34d2a9e3baadb141a17c77ef3c565dd5ee5054b91cf63bb3c6ab83f72ab3aafe93d0fc3c2348b764fafb0b1c0733de51459aeab46580384bf9d74c4e28164b7cde247f891ba07891c9d872ad2bb")]
+
+namespace Microsoft.AspNetCore.Cryptography
```
[the remark this answers is not in the document]
Comment (JunTaoLuo) [reply src/DataProtection/Cryptography.Internal/ref/Microsoft.AspNetCore.Cryptography.Internal.Manual.cs]: I have a list of issues I ran into while doing this. I'll open an issue later.

[the remark this answers is not in the document]
Comment (dougbu) [reply src/DataProtection/Cryptography.Internal/ref/Microsoft.AspNetCore.Cryptography.Internal.Manual.cs]: 👍

File: src/SignalR/common/Shared/PipeWriterStream.cs (hunk 9 above)
Comment (dougbu) [reply src/SignalR/common/Shared/PipeWriterStream.cs]: > the internal APIs are not generated by default so this using must be removed

Is this a change in the build? Were warnings and errors suppressed somehow before your changes?

File: src/Mvc/Mvc.NewtonsoftJson/src/Microsoft.AspNetCore.Mvc.NewtonsoftJson.csproj (hunk 6 above)
Comment (dougbu) [reply src/Mvc/Mvc.NewtonsoftJson/src/Microsoft.AspNetCore.Mvc.NewtonsoftJson.csproj]: What is the answer to my question / resolution here?

Comment (dougbu) [reply src/Mvc/Mvc.NewtonsoftJson/src/Microsoft.AspNetCore.Mvc.NewtonsoftJson.csproj]: Oh, I see. File no longer changed.

File: eng/targets/ReferenceAssembly.targets [hunk 12]
```diff
@@ -59,7 +59,7 @@
     <PropertyGroup>
       <_GenAPICommand Condition="'$(MSBuildRuntimeType)' == 'core'">"$(DotNetTool)" --roll-forward-on-no-candidate-fx 2 "$(_GenAPIPath)"</_GenAPICommand>
       <_GenAPICmd>$(_GenAPICommand)</_GenAPICmd>
-      <_GenAPICmd>$(_GenAPICmd) "$(TargetPath)"</_GenAPICmd>
+      <_GenAPICmd>$(_GenAPICmd) "$(TargetPath)" $(AdditionalGenApiCmdOptions)</_GenAPICmd>
```
[the remark this answers is not in the document]
Comment (JunTaoLuo) [reply eng/targets/ReferenceAssembly.targets]: I was using it pretty regularly when fixing our internal APIs. I could remove this at the end of this PR when I won't need to use it anymore.

File: src/SignalR/common/Shared/PipeWriterStream.cs (hunk 9 above)
Comment (JunTaoLuo) [reply src/SignalR/common/Shared/PipeWriterStream.cs]: So the history of this change is:
1. ValueTaskExtensions was added in https://github.com/aspnet/AspNetCore/commit/9de42d516e6ee9c8eb17700e134f3ac8a7a682ff to the .Internal namespace
2. This was removed later when the extension method was no longer needed but the using statement was not removed.
3. The Microsoft.AspNetCore.Internal namespace still existed since other internal APIs from unrelated transitive references still contained this namespace. 
4. Changing to ref assemblies meant that this namespace no longer exists for most cases. In this case, no internal APIs were being used so I'm removing the using statements.

File: eng/targets/ResolveReferences.targets [hunk 13]
```diff
@@ -50,10 +50,10 @@
     <UseProjectReferences Condition=" '$(UseProjectReferences)' == '' AND '$(IsImplementationProject)' != 'true' ">true</UseProjectReferences>
     <UseProjectReferences Condition=" '$(UseProjectReferences)' == '' ">false</UseProjectReferences>
 
-    <ReferenceReferenceAssemblies Condition=" '$(ReferenceReferenceAssemblies)' == '' AND '$(IsReferenceAssemblyProject)' == 'true'">true</ReferenceReferenceAssemblies>
+    <ReferenceReferenceAssemblies Condition=" '$(ReferenceReferenceAssemblies)' == '' AND ('$(IsReferenceAssemblyProject)' == 'true' OR '$(IsImplementationProject)' == 'true')">true</ReferenceReferenceAssemblies>
```
[the remark this answers is not in the document]
Comment (JunTaoLuo) [reply eng/targets/ResolveReferences.targets]: I think we need additional logic to handle it. Haven't gotten to it.

**Comment by JunTaoLuo** [comment]
Closed in favour of https://github.com/aspnet/AspNetCore/pull/14425
